e2e Slack notifications: update file upload logic

jeherve opened the pull request:
## Proposed changes:

Slack will soon deprecate the files.upload API:
https://docs.slack.dev/changelog/2024-04-a-better-way-to-upload-files-is-here-to-stay/

Let's move to `@slack/web-api` and its `filesUploadV2` instead.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion

* N/A

## Does this pull request change what data or activity we track or use?
* No

## Testing instructions:

* Do the tests pass?
* Check the notification generated here: p1745344777214659/1745344773.842519-slack-C1Y1F7ZM1

jeherve committed:
e2e Slack notifications: update file upload logic

Slack will soon deprecate the files.upload API:
https://docs.slack.dev/changelog/2024-04-a-better-way-to-upload-files-is-here-to-stay/

We are encouraged to move to files.getUploadURLExternal and files.completeUploadExternal instead.
https://api.slack.com/methods/files.getUploadURLExternal
https://api.slack.com/methods/files.completeUploadExternal

jeherve committed:
changelog

anomiex commented on projects/github-actions/test-results-to-slack/changelog/update-slack-notifications-deprecation:
```diff
@@ -0,0 +1,4 @@
+Significance: minor
+Type: changed
+
+Slack API: update file upload logicin preparation for files.upload deprecation
```
```suggestion
Slack API: update file upload logic in preparation for files.upload deprecation
```

jeherve committed:
Expand the tests to avoid coverage failure

anomiex commented on projects/github-actions/test-results-to-slack/src/slack.js:
```diff
@@ -156,6 +156,70 @@ async function getMessage( client, channelId, identifier ) {
+		}
+
+		// Step 2: Upload file to the URL
+		const fileStream = fs.createReadStream( filePath );
+		const fileBuffer = await new Promise( ( resolve, reject ) => {
+			const chunks = [];
+			fileStream.on( 'data', chunk => chunks.push( chunk ) );
+			fileStream.on( 'end', () => resolve( Buffer.concat( chunks ) ) );
+			fileStream.on( 'error', reject );
+		} );
+
+		const uploadResponse = await fetch( upload_url, {
+			method: 'POST',
+			body: fileBuffer,
+			headers: {
+				'Content-Type': 'application/octet-stream',
+			},
+		} );
```
Looks like, instead of reading the stream into memory, it may be possible to do like this
```suggestion
		const fileStream = fs.createReadStream( filePath );
		const uploadResponse = await fetch( upload_url, {
			method: 'POST',
			body: fileStream,
			duplex: 'half',
			headers: {
				'Content-Type': 'application/octet-stream',
			},
		} );
```
Or, if that doesn't work or you'd rather have a Buffer anyway, you should be able to read it directly like
```suggestion
		const fileBuffer = await fs.promises.readFile( filePath );
		const uploadResponse = await fetch( upload_url, {
			method: 'POST',
			body: fileBuffer,
			headers: {
				'Content-Type': 'application/octet-stream',
			},
		} );
```

anomiex reviewed (commented):
This seems reasonable, as far as it goes. Left some suggestions inline.

OTOH, I see there's mention that there should be a `client.filesUploadV2()` method. Is there a reason not to use that instead of rolling our own?

jeherve committed:
changelog: update wording.

Co-authored-by: Brad Jorsch <anomiex@users.noreply.github.com>

jeherve commented:
> OTOH, I see there's mention that there should be a `client.filesUploadV2()` method. Is there a reason not to use that instead of rolling our own?

Oh, I missed that option, I didn't pay attention enough to that part of the docs, I focussed on `[files.getUploadURLExternal](https://api.slack.com/methods/files.getUploadURLExternal)` and `[files.completeUploadExternal](https://api.slack.com/methods/files.completeUploadExternal)` instead.

Let's use that instead!

jeherve committed:
Update implementation to use @slack/web-api

See https://github.com/Automattic/jetpack/pull/43196#pullrequestreview-2784732826

anomiex commented on projects/github-actions/test-results-to-slack/src/slack.js:
```diff
@@ -156,6 +156,41 @@ async function getMessage( client, channelId, identifier ) {
+			throw new Error( `File not found: ${ filePath }` );
+		}
+
+		const response = await client.filesUploadV2( {
+			channel_id,
+			thread_ts,
+			file: filePath,
+			filename,
+		} );
```
I wonder why you don't just inline this above like the previous code did.

The file existence check on line 172 is redundant to the one on line 27.

jeherve replied:
Yes, now that I got rid of the extra complexity and use filesUploadV2, we can revert back to what it was. 👍 Done in fdea35ffa8cf7e88861cb85e251c07c68720763d

anomiex commented on projects/github-actions/test-results-to-slack/src/slack.js:
```diff
@@ -52,6 +51,7 @@ async function postOrUpdateMessage( client, update, options ) {
 				} );
```
If you're adding error checking everywhere else, why not check `response.ok` here too?

jeherve replied:
Good point. Done in 541249a59607c48c03a281d53e49ee6bf13e9e98

anomiex reviewed (commented):
The upload change looks reasonable. Haven't tried to test it though.

I note there's an unmentioned behavior change in here. Previously, if one file or one channel failed, it would ignore the errors and continue posting any following chunks and posting to [any additional channels](https://github.com/Automattic/jetpack/blob/bf38b5fdcfed50547db4919ad5075f012d00a72c/projects/github-actions/test-results-to-slack/src/index.js#L35-L37). Now you're rethrowing errors in all the code paths, so the first failure will prevent anything else from being sent to Slack. That should probably be mentioned.

jeherve committed:
Additional error check

See https://github.com/Automattic/jetpack/pull/43196/files#r2054583550

Co-authored-by: Brad Jorsch <anomiex@users.noreply.github.com>

jeherve committed:
Bring filesUploadV2 back into postOrUpdateMessage

We don't need to extract things anymore, since we simplified the upload in bf38b5fdcfed50547db4919ad5075f012d00a72c

Co-authored-by: Brad Jorsch <anomiex@users.noreply.github.com>

jeherve commented:
> I note there's an unmentioned behavior change in here. Previously, if one file or one channel failed, it would ignore the errors and continue posting any following chunks and posting to [any additional channels](https://github.com/Automattic/jetpack/blob/bf38b5fdcfed50547db4919ad5075f012d00a72c/projects/github-actions/test-results-to-slack/src/index.js#L35-L37). Now you're rethrowing errors in all the code paths, so the first failure will prevent anything else from being sent to Slack. That should probably be mentioned.

👍 
This was mostly to add more tests to please the coverage test. Maybe I can just revert things now.

jeherve committed:
Merge remote-tracking branch 'origin/trunk' into update/slack-notifications-deprecation

jeherve committed:
Update test

jeherve committed:
Remove extra handling of errors

anomiex reviewed (approved):
Looks reasonable. Haven't tested.

jeherve commented:
I tested it with this message, it seems to work well:
p1745414533797859/1745344773.842519-slack-C1Y1F7ZM1

Merging.
